src/Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -405,15 +416,37 @@ public void ValidateBlock(ContextInformation context)
                 // Calculate the consensus flags and check they are valid.
                 context.Flags = this.NodeDeployments.GetFlags(context.BlockValidationContext.ChainedBlock);
 
-                int lastCheckpointHeight = this.checkpoints.GetLastCheckpointHeight();
-                if (context.BlockValidationContext.ChainedBlock.Height > lastCheckpointHeight)
+                // Check whether to use checkpoint to skip block validation.
+                context.BlockValidationContext.SkipValidation = false; 
+                if (this.nodeSettings.Consensus.UseCheckpoints)
+                {
+                    int lastCheckpointHeight = this.checkpoints.GetLastCheckpointHeight();
+                    context.BlockValidationContext.SkipValidation = context.BlockValidationContext.ChainedBlock.Height <= lastCheckpointHeight;
+                    if (context.BlockValidationContext.SkipValidation)
+                        this.logger.LogTrace("Block validation will be partially skipped due to block height {0} is not greater than last checkpointed block height {1}.", context.BlockValidationContext.ChainedBlock.Height, lastCheckpointHeight);
+                }
+
+                // Check whether to use assumevalid switch to skip validation.
+                if (!context.BlockValidationContext.SkipValidation && (this.nodeSettings.Consensus.BlockAssumedValid != null))
+                {
+                    ChainedBlock assumeValidBlock = this.Chain.GetBlock(this.nodeSettings.Consensus.BlockAssumedValid);
+                    if (assumeValidBlock != null)
+                    {
+                        context.BlockValidationContext.SkipValidation = context.BlockValidationContext.ChainedBlock.Height <= assumeValidBlock.Height;
+                        if (context.BlockValidationContext.SkipValidation)
+                            this.logger.LogTrace("Block validation will be partially skipped due to block height {0} is not greater than assumed valid block height {1}.", context.BlockValidationContext.ChainedBlock.Height, assumeValidBlock.Height);
+                    }
+                    else this.logger.LogWarning("Cannot find block specified by config flag assumevalid: '{0}'. Validation will not be skipped.", this.nodeSettings.Consensus.BlockAssumedValid);
```

Comment:
I still think the else needs to throw.
If the assume valid is not in the main chain do we continue with validation enabled or stop?
If the assume valid hash was passed in configuation then the user wants to be on a certain chain.
If we do decide to throw the block should not be added to the InvalidBlocks received.

Reply:
Core is just ignoring flag and doing validation if it can't find the hash. If we want to throw an exception I think we might want to validate it earlier (maybe after headers are downloaded) as with the  default settings this code won't even get executed until after the checkpoints. So you won't see it until near the end of the IBD.

Reply:
Good point about validating earlier ofcourse.
Right if that's what core are doing then we do the same.

TBH I think this code will replace checkpoints, this is duplicate effort of the same thing only AssumeValid is more flexible.

Reply:
@dangershony if the assume valid hash is not on your chain you need to do full validation and this is perfectly valid case - it just mean that the chain that you knew that was valid was reorged and that is perfectly fine thing to happen, this is the difference between this feature and checkpoints